aexpect/ops.py

```diff
@@ -0,0 +1,439 @@
+    """
+    Calculate hash of given file from a session.
+
+    :param session: session to run the command on
+    :type session: ShellSession
+    :param str filename: full path of file that should be hashed
+    :param str method: method for hashing
+    :returns: hash of file, a hex number
+    :rtype: str
+    :raises: :py:class:`RuntimeError` if hash command failed (e.g. file not
+             found) or resulting output does not have expected format
+    """
+    cmd = f"{method}sum"
+    expect_len = 32
+
+    # run cmd on shell
+    status, output = session.cmd_status_output(cmd + ' ' + quote(filename))
+    if output:
+        output = output.strip()
+    if status != 0 or not output:
+        raise RuntimeError(f'Could not hash {filename} using {cmd}: {output}')
+
+    # parse output
+    hash_str = output.split(maxsplit=1)[0].lower()
+
+    # check length and all chars are hex
+    if expect_len and len(hash_str) != expect_len:
+        raise RuntimeError(f'Resulting hash string has unexpected length '
+                           f'{len(hash_str)}: {hash_str}')
+    if hash_str.strip('0123456789abcdef'):
+        raise RuntimeError('Resulting hash string has unexpected characters: '
+                           + hash_str)
+    return hash_str
+
+
+def extract_tarball(session, tarball, target_dir):
+    """
+    Extract tarball to given dir from a session.
+
+    :param session: session to run the command on
+    :type session: ShellSession
+    :param str tarball: full path of tar file that should be extracted
+    :param str target_dir: name of directory where tarball should be extracted
+    :raises: :py:class:`RuntimeError` if tar command returned non-null
+    """
+    cmd = f'tar -C {quote(target_dir)} --strip-components=1 -xapf {quote(tarball)}'
+    status, output = session.cmd_status_output(cmd)
+    if output:
+        output = output.strip()
+    if status != 0:
+        raise RuntimeError(f'Failed to extract {tarball} to {target_dir}: {output}')
+
+
+def ls(session, path, is_glob=False, flags="-1UNq"):  # pylint: disable=C0103
+    """
+    Run `ls` in given directory through a session.
+
+    :param session: session to run the command on
+    :type session: ShellSession
+    :param str path: absolute or relative path to list
+    :param bool is_glob: whether to expand or not the path as a glob pattern
+    :returns: names of files in dir (without path components)
+    :rtype: [str]
+
+    Just like :py:func:`os.listdir`, does not include file names starting with
+    dot (`'.'`)
+    """
+    cmd = f'ls {flags} {path}' if is_glob else f'ls {flags} {quote(path)}'
+    status, output = session.cmd_status_output(cmd)
+    if output:
+        output = output.strip()
+    if status == 2:      # probably just nothing found
+        return []
+    if status != 0:
+        raise RuntimeError(f'Failed to ls {path}: {output}')
+    return output.splitlines()
+
+
+def cat(session, filename):
```

Comment:
The `flags` could be useful here as well (eg. for `-v`) but it's not must-have in this case.

Reply:
Alright but better to just add to the most critical cases right now since later refactoring could easily make flags available for every command and this might duplicate the effort.